### PR TITLE
Validate protected command buffer and filter cubic, and some refactors

### PR DESCRIPTION
**Pull request by locke-lunarg**
> VUID-vkCmdDraw-commandBuffer-02707
> If commandBuffer is an unprotected command buffer, any resource accessed by the VkPipeline object bound to the pipeline bind point used by this command must not be a protected resource

Validate the image and buffer in DescriptorSet. If the commandbuffer is unprotected, they MUST be unprotected.

> VUID-vkCmdTraceRaysNV-commandBuffer-02712
> If commandBuffer is a protected command buffer, any resource written to by the VkPipeline object bound to the pipeline bind point used by this command must not be an unprotected resource

Validate the image and buffer in DescriptorSet. If the commandbuffer is protected, they MUST be protected or only-read unprotected.

> VUID-vkCmdDraw-filterCubic-02694
> Any VkImageView being sampled with VK_FILTER_CUBIC_EXT as a result of this command must have a VkImageViewType and format that supports cubic filtering, as specified by VkFilterCubicImageViewImageFormatPropertiesEXT::filterCubic returned by vkGetPhysicalDeviceImageFormatProperties2

In PostCallRecordCreateImageView, it saves the result of VkFilterCubicImageViewImageFormatPropertiesEXT in IMAGE_VIEW_STATE, and then validate both sampler and image descriptors. If sampler is VK_FILTER_CUBIC_EXT, the image view needs to support filterCubic.

> VUID-vkCmdDraw-filterCubicMinmax-02695
> Any VkImageView being sampled with VK_FILTER_CUBIC_EXT with a reduction mode of either VK_SAMPLER_REDUCTION_MODE_MIN or VK_SAMPLER_REDUCTION_MODE_MAX as a result of this command must have a VkImageViewType and format that supports cubic filtering together with minmax filtering, as specified by VkFilterCubicImageViewImageFormatPropertiesEXT::filterCubicMinmax returned by vkGetPhysicalDeviceImageFormatProperties2

Validate both sampler and image descriptors. If sampler is VK_FILTER_CUBIC_EXT and either VK_SAMPLER_REDUCTION_MODE_MIN or VK_SAMPLER_REDUCTION_MODE_MAX , the image view needs to support filterCubicMinmax.

> VUID-vkCmdDraw-maxMultiviewInstanceIndex-02688
> If the draw is recorded in a render pass instance with multiview enabled, the maximum instance index must be less than or equal to VkPhysicalDeviceMultiviewProperties::maxMultiviewInstanceIndex

Save VkPhysicalDeviceMultiviewFeatures in PostCallRecordCreateDevice. Validate if commandbuffer is bind with a renderpass, and then if multiview feature is true, and then (instanceCount + firstInstance) MUST be eqaul or less than maxMultiviewInstanceIndex. Some Cmd's instance value is in VkBuffer. Skip those Cmd.

### CODE REVIEW

**Comment by locke-lunarg** [comment]
Some refactors.

1) Moved looking for sampler descriptor from ValidateDescriptorSetBindingData into PreCallRecordCmdBindPipeline & PreCallRecordCmdBindDescriptorSets to impre

2) https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/2226#pullrequestreview-513191074
   1) Change map BindingReqMap into unordered_map
   2) Change map lastBoundnge into array

3) Moved some validation for both image and sampler descriptor from` if (descriptor_class == DescriptorClass::ImageSampler || descriptor_class == DescriptorClass::PlainSampler)` into `if (descriptor_class == DescriptorClass::ImageSampler || descriptor_class == DescriptorClass::Image)`.

   ` if (descriptor_class == DescriptorClass::ImageSampler || descriptor_class == DescriptorClass::PlainSampler)` only validates sampler descriptor.

**Comment by sfricke-samsung** [comment]
So first off you are attempting to finish the last 2 sets of Protected memory validation which will close #1996

So  "resource" can be anything that can be written to which is not limited to descriptors. For draws calls you will also want to check for Framebuffer attachments as well

**Comment by locke-lunarg** [comment]
@sfricke-samsung  Thanks. Do you think it should also include binding vertices and indices?

**Comment by sfricke-samsung** [comment]
@locke-lunarg  Ok, so re-read the VU

`02712` is "any resource written"

and

`02707` is "any resource accessed"

The "accessed" is defined as "store, load, atomic" else where in the spec

So yes, vertex input would be "Accessed" and possibly the use of Physical Storage Buffer as well

basically, I put these off as they are not a trivial addition of a VU

**Comment by locke-lunarg** [comment]
@sfricke-samsung  Added validating framebuffer's attachments and binding vertex & index buffer.

**Review by Tony-LunarG — CHANGES_REQUESTED** [review]

File: layers/core_validation_types.h [hunk 1]
```diff
@@ -1209,6 +1215,40 @@ typedef ImageSubresourceLayoutMap::LayoutMap GlobalImageLayoutRangeMap;
 typedef std::unordered_map<VkImage, std::unique_ptr<GlobalImageLayoutRangeMap>> GlobalImageLayoutMap;
 typedef std::unordered_map<VkImage, std::unique_ptr<ImageSubresourceLayoutMap>> CommandBufferImageLayoutMap;
 
+enum LvlBindPoint {
```
Comment (jzulauf-lunarg) [inline layers/core_validation_types.h:1218]: Naming issue no _

Computer -> Compute

File: layers/core_validation_types.h [hunk 2]
```diff
@@ -1209,6 +1215,40 @@ typedef ImageSubresourceLayoutMap::LayoutMap GlobalImageLayoutRangeMap;
...
+    BindPoint_Graphics = 0,
+    BindPoint_Computer = 1,
+    BindPoint_Ray_Tracing = 2,
+};
+
+static VkPipelineBindPoint inline ConvertToPipelineBindPoint(LvlBindPoint bind_point) {
+    switch (bind_point) {
```
Comment (jzulauf-lunarg) [inline layers/core_validation_types.h:1225]: Given that graphics and compute are the same and only ray tracing differs... is there a simpler form... (look up, force cast, something.. for the first two...)

File: layers/core_validation_types.h [hunk 3]
```diff
@@ -1237,10 +1277,9 @@ struct CMD_BUFFER_STATE : public BASE_NODE {
     //  long-term may want to create caches of "lastBound" states and could have
     //  each individual CMD_NODE referencing its own "lastBound" state
     // Store last bound state for Gfx & Compute pipeline bind points
-    std::map<uint32_t, LAST_BOUND_STATE> lastBound;
+    std::array<LAST_BOUND_STATE, 3> lastBound;  // index is LvlBindPoint.
```
Comment (jzulauf-lunarg) [inline layers/core_validation_types.h:1280]: Define a LvlBindPoint BindPointSize (or Count) and use that as the array size.

**Review by jzulauf-lunarg — COMMENTED** [review]
didn't do a full review... just a couple quick notes

File: layers/core_validation.cpp [hunk 4]
```diff
@@ -1081,6 +1096,48 @@ bool CoreChecks::ValidateCmdBufDrawState(const CMD_BUFFER_STATE *cb_node, CMD_TY
         if (cb_node->activeRenderPass && cb_node->activeFramebuffer) {
             const auto &subpass = cb_node->activeRenderPass->createInfo.pSubpasses[cb_node->activeSubpass];
             attachment_views = cb_node->activeFramebuffer->GetUsedAttachments(subpass, cb_node->imagelessFramebufferAttachments);
+
+            // Verify attachments for unprotected/protected command buffer.
+            if (subpass.colorAttachmentCount > 0 && subpass.pColorAttachments) {
+                for (uint32_t i = 0; i < subpass.colorAttachmentCount; ++i) {
+                    uint32_t index = subpass.pColorAttachments[i].attachment;
```
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply layers/core_validation.cpp]: In FRAMEBUFFER_STATE::GetUsedAttachments, it has checked VK_ATTACHMENT_UNUSED, if it's VK_ATTACHMENT_UNUSED, the view is VK_NULL_HANDLE

File: layers/core_validation.cpp [hunk 5]
```diff
@@ -725,6 +725,24 @@ bool CoreChecks::ValidatePipelineDrawtimeState(const LAST_BOUND_STATE &state, co
...
+            if (buf_state) {
+                skip |=
+                    ValidateProtectedBuffer(pCB, buf_state, caller, vuid.unprotected_command_buffer, "Buffer is a binding vertex");
+            }
+        }
+    }
+    if (pCB->index_buffer_binding.buffer != VK_NULL_HANDLE) {
+        const auto *buf_state = Get<BUFFER_STATE>(pCB->index_buffer_binding.buffer);
+        if (buf_state) {
+            skip |= ValidateProtectedBuffer(pCB, buf_state, caller, vuid.unprotected_command_buffer, "Buffer is a binding index");
```
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply layers/core_validation.cpp:742]: Fixed

File: layers/core_validation.cpp [hunk 6]
```diff
@@ -725,6 +725,24 @@ bool CoreChecks::ValidatePipelineDrawtimeState(const LAST_BOUND_STATE &state, co
...
     const DrawDispatchVuid vuid = GetDrawDispatchVuid(cmd_type);
 
+    // Verify vertex & index buffer for unprotected command buffer.
+    // Because vertex & index buffer is read only, it doesn't need to care protected command buffer case.
+    for (const auto &buffer_binding : current_vtx_bfr_binding_info) {
+        if (buffer_binding.buffer != VK_NULL_HANDLE) {
+            const auto *buf_state = Get<BUFFER_STATE>(buffer_binding.buffer);
+            if (buf_state) {
+                skip |=
+                    ValidateProtectedBuffer(pCB, buf_state, caller, vuid.unprotected_command_buffer, "Buffer is a binding vertex");
```
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply layers/core_validation.cpp:735]: fixed

File: layers/core_validation.cpp [hunk 7]
```diff
@@ -1081,6 +1096,48 @@ bool CoreChecks::ValidateCmdBufDrawState(const CMD_BUFFER_STATE *cb_node, CMD_TY
...
+                            result |= ValidateUnprotectedImage(cb_node, view_state->image_state.get(), function,
+                                                               vuid.protected_command_buffer, "Image is a color attachment.");
+                            result |= ValidateProtectedImage(cb_node, view_state->image_state.get(), function,
+                                                             vuid.unprotected_command_buffer, "Image is a color attachment.");
+                        }
+                    }
+                }
+            }
+            if (subpass.pDepthStencilAttachment) {
+                uint32_t index = subpass.pDepthStencilAttachment->attachment;
```
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply layers/core_validation.cpp]: In FRAMEBUFFER_STATE::GetUsedAttachments, it has checked VK_ATTACHMENT_UNUSED, if it's VK_ATTACHMENT_UNUSED, the view is VK_NULL_HANDLE

File: layers/core_validation.cpp [hunk 8]
```diff
@@ -1081,6 +1096,48 @@ bool CoreChecks::ValidateCmdBufDrawState(const CMD_BUFFER_STATE *cb_node, CMD_TY
...
+                        result |= ValidateProtectedImage(cb_node, view_state->image_state.get(), function,
+                                                         vuid.unprotected_command_buffer, "Image is a depthstencil attachment.");
+                    }
+                }
+            }
+
+            // Because inputAttachment is read only, it doesn't need to care protected command buffer case.
+            if (subpass.inputAttachmentCount > 0 && subpass.pInputAttachments) {
+                for (uint32_t i = 0; i < subpass.inputAttachmentCount; ++i) {
+                    uint32_t index = subpass.pInputAttachments[i].attachment;
```
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply layers/core_validation.cpp]: In FRAMEBUFFER_STATE::GetUsedAttachments, it has checked VK_ATTACHMENT_UNUSED, if it's VK_ATTACHMENT_UNUSED, the view is VK_NULL_HANDLE

File: layers/core_validation_types.h [hunk 9]
```diff
@@ -1236,10 +1276,9 @@ struct CMD_BUFFER_STATE : public BASE_NODE {
...
```
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply layers/core_validation_types.h:1280]: I cannot find out lastBoundnge

File: layers/core_validation_types.h [hunk 10]
```diff
@@ -178,12 +178,15 @@ struct less<SamplerUsedByImage> {
...
 }  // namespace std
 
+struct SAMPLER_STATE;
 struct DescriptorRequirement {
     descriptor_req reqs;
-    std::vector<std::map<SamplerUsedByImage, std::map<VkDescriptorSet, const cvdescriptorset::Descriptor *>>>
-        samplers_used_by_image;  // Copy from StageState.interface_var. BUT it combines from plural shader stages.
+    bool is_writable;
+    std::vector<std::map<SamplerUsedByImage, const cvdescriptorset::Descriptor *>>
+        samplers_used_by_image;  // Copy from StageStCmdDrawDispatchInfoate.interface_var. It combines from plural shader stages.
```
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply layers/core_validation_types.h]: fixed

File: layers/core_validation_types.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (Tony-LunarG) [reply layers/core_validation_types.h:1280]: In the commit message "layers: Change map lastBoundnge into array"

File: layers/core_validation_types.h (hunk 1 above)
Comment (locke-lunarg) [reply layers/core_validation_types.h:1218]: fixed

File: layers/core_validation_types.h (hunk 3 above)
Comment (locke-lunarg) [reply layers/core_validation_types.h:1280]: added

File: layers/core_validation_types.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (locke-lunarg) [reply layers/core_validation_types.h]: Thanks. Fixed

File: layers/core_validation_types.h (hunk 2 above)
Comment (locke-lunarg) [reply layers/core_validation_types.h:1226]: fixed

File: layers/core_validation.h [hunk 11]
```diff
@@ -62,6 +62,11 @@ struct DrawDispatchVuid {
     const char* sampler_bias_offset;
     const char* vertex_binding_attribute;
     const char* dynamic_state_setting_commands;
+    const char* unprotected_command_buffer;
+    const char* protected_command_buffer;
+    const char* maxMultiviewInstanceIndex;  // TODO: Some instance values are in VkBuffer.The validation in those Cmds is skipped.
```
Comment (mark-lunarg) [inline layers/core_validation.h:67]: All three of these should be lower_snake_case, please (like the others).

Comment (locke-lunarg) [reply layers/core_validation.h]: fixed

Comment (mark-lunarg) [reply layers/core_validation.h]: Thanks!  I have a PR up that refactors these `DrawDispatchVuid `structures, but I'll wait for your PR so you won't need to deal with the rebase.

Comment (locke-lunarg) [reply layers/core_validation.h]: Thanks!

**Comment by locke-lunarg** [comment]
@Tony-LunarG @mark-lunarg @jzulauf-lunarg  Refactored the attachments part to ensure their lifetime during commandbuffer. It's ready to review.

**Review by Tony-LunarG — APPROVED** [review]
LGTM - thanks!

**Pull request merged**